paddle/fluid/platform/place.h

```diff
@@ -65,6 +65,18 @@ bool is_cpu_place(const Place &);
 bool places_are_same_class(const Place &, const Place &);
 bool is_same_place(const Place &, const Place &);
 
+struct PlaceHash {
+  std::size_t operator()(const Place &p) const {
+    constexpr size_t num_dev_bits = 4;
```

Comment:
4 bit is not enough, the GPU box product has 32 cards in one node. Then will lead to an overlap of `dev_id << num_dev_bits | p.which()`